flang/include/flang/Optimizer/Dialect/Coarray/CoarrayOps.td

```diff
@@ -0,0 +1,828 @@
+      - `atom` and `value`: Shall be a coarray of type integer with kind ATOMIC_INT_KIND 
+        from module ISO_FORTRAN_ENV.
+  }];
+
+  let arguments = (ins AnyIntegerType:$image_num,
+                       fir_BoxType:$atom,
+                       Arg<fir_BoxType, "", [MemRead]>:$value,
+                       Arg<Optional<AnyRefOrBoxType>, "", [MemWrite]>:$errmsg);
+  let results = (outs AnyIntegerType:$stat);
+}
+
+def coarray_AtomicCasOp : coarray_Op<"atomic_cas", []> {
+  let summary = "Atomic compare and swap";
+  let description = [{
+    Arguments: 
+      - `image_num`: Shall be an integer that correspond to the image index of 
+        `atom` where the operation need to be performed. 
+      - `atom`, `old`, `new` and `compared`: Shall be a coarray of type integer with kind
+         ATOMIC_INT_KIND or ATOMIC_LOGICAL_KIND from module ISO_FORTRAN_ENV. They need to 
+         have the same type and kind.
+  }];
+
+  let arguments = (ins AnyIntegerType:$image_num,
+                       fir_BoxType:$atom,
+                       Arg<fir_BoxType, "", [MemWrite]>:$old,
+                       Arg<fir_BoxType, "", [MemRead]>:$compare,
+                       Arg<fir_BoxType, "", [MemRead]>:$new_val,
+                       Arg<Optional<AnyRefOrBoxType>, "", [MemWrite]>:$errmsg);
+  let results = (outs AnyIntegerType:$stat);
+}
+
+def coarray_AtomicDefineOp : coarray_Op<"atomic_define", []> {
+  let summary = "Define a variable atomically";
+  let description = [{
+    Arguments: 
+      - `image_num`: Shall be an integer that correspond to the image index of 
+        `atom` where the operation need to be performed. 
+      - `atom` and `value`: Shall be a coarray of type integer with kind ATOMIC_INT_KIND 
+        or ATOMIC_LOGICAL_KIND from module ISO_FORTRAN_ENV.
+  }];
+
+  let arguments = (ins AnyIntegerType:$image_num,
+                       fir_BoxType:$atom,
+                       Arg<fir_BoxType, "", [MemRead]>:$value,
+                       Arg<Optional<AnyRefOrBoxType>, "", [MemWrite]>:$errmsg);
+  let results = (outs AnyIntegerType:$stat);
+}
+
+def coarray_AtomicRefOp : coarray_Op<"atomic_ref", []> {
+  let summary = "Reference a variable atomically";
+  let description = [{
+    Arguments: 
+      - `image_num`: Shall be an integer that correspond to the image index of 
+        `atom` where the operation need to be performed. 
+      - `atom` and `value`: Shall be a coarray of type integer with kind ATOMIC_INT_KIND 
+        or ATOMIC_LOGICAL_KIND from module ISO_FORTRAN_ENV.
+  }];
+
+  let arguments = (ins AnyIntegerType:$image_num,
+                       fir_BoxType:$atom,
+                       Arg<fir_BoxType, "", [MemRead]>:$value,
+                       Arg<Optional<AnyRefOrBoxType>, "", [MemWrite]>:$errmsg);
+  let results = (outs AnyIntegerType:$stat);
+}
+
+//===----------------------------------------------------------------------===//
+// Collective Operations
```

Comment:
Do Collective Subroutines really belong in this dialect?    

These are often conflated during informal conversations, but Collective Subroutines have absolutely nothing to do with coarrays. They are a completely orthogonal multi-image feature. In fact the argument to collective subroutines is even prohibited from being a coindexed object.    

I'm not sure these belong in ANY dialect - for most static purposes, collective subroutines are just a regular intrinsic subroutine. IMHO they don't belong lumped into a coarray dialect any more than say, the SQRT intrinsic.